# Scripts for the manuscript '*Studying the genetics of participation using footprints left on the ascertained genotypes*'
Scripts to reproduce the analyses in Benonisdottir, S., Kong, A. '*Studying the genetics of participation using footprints left on the ascertained genotypes*'. Nat Genet (2023). https://www.nature.com/articles/s41588-023-01439-2 <br />
* **/GenCorrH2**: Performing LD score regression with the program [LDSC version 1.0.1](https://github.com/bulik/ldsc) 
to attain LD score regression intercepts, genetic correlation estimates and inputs for heritability estimates on a liability scale - *Table 2*, *Supplementary Table 2* and *Supplementary Table 3*.
* **/PGSPheno**: Performing linear and logistic regression to examine the relationship between the primary participation polygenic score 
and various phenotypes. - *Table1*, *Figure 3* and *Supplementary Table 1*.
* **/PartGWAS**: Performing primary participation GWAS. <br />
* **/Simulations**: Simulations to investigate the relative allele frequencies in different groups of individuals (participants, non-participants, individuals with a participating/non-participating sibling) 
and genetic segments (IBD shared and not-shared) - *Figure 4* and *Extended Data Figure 8*. 
* **/SumStatAnalysis**: Computing T-statistics from the summary statistics files generated with scripts in /PartGWAS and investigating 
their relationship with allele frequency - *Extended Data Figure 5*, *Extended Data Figure 6* and *Extended Data Figure 7*. <br />
